Mathlib/Topology/Order/LowerUpperTopology.lean

```diff
@@ -264,7 +444,24 @@ instance [Preorder α] [TopologicalSpace α] [LowerTopology α] [OrderBot α] [P
     · exact GenerateOpen.basic _ ⟨(a, ⊥), by simp [Ici_prod_eq, prod_univ]⟩
     · exact GenerateOpen.basic _ ⟨(⊥, b), by simp [Ici_prod_eq, univ_prod]⟩
 
-section CompleteLattice
+instance [Preorder α] [TopologicalSpace α] [UpperTopology α] [OrderTop α] [Preorder β]
+    [TopologicalSpace β] [UpperTopology β] [OrderTop β] : UpperTopology (α × β) where
+  topology_eq_upperTopology := by
+    refine' le_antisymm (le_generateFrom _) _
+    · rintro _ ⟨x, rfl⟩
+      exact ((UpperTopology.isClosed_Iic _).prod <| UpperTopology.isClosed_Iic _).isOpen_compl
+    rw [(UpperTopology.isTopologicalBasis.prod UpperTopology.isTopologicalBasis).eq_generateFrom,
+      le_generateFrom_iff_subset_isOpen, image2_subset_iff]
+    rintro _ ⟨s, hs, rfl⟩ _ ⟨t, ht, rfl⟩
+    dsimp
+    simp_rw [coe_lowerClosure, compl_iUnion, prod_eq, preimage_iInter, preimage_compl]
+    -- without `let`, `refine` tries to use the product topology and fails
+    let _ : TopologicalSpace (α × β) := generateFrom { s | ∃ a, (Iic a)ᶜ = s }
+    refine (isOpen_biInter hs fun a _ => ?_).inter (isOpen_biInter ht fun b _ => ?_)
+    · exact GenerateOpen.basic _ ⟨(a, ⊤), by simp [Iic_prod_eq, prod_univ]⟩
+    · exact GenerateOpen.basic _ ⟨(⊤, b), by simp [Iic_prod_eq, univ_prod]⟩
```

Comment:
```suggestion
  topology_eq_upperTopology := by
    suffices : LowerTopology (α × β)ᵒᵈ
    · exact LowerTopology.topology_eq_lowerTopology (α := (α × β)ᵒᵈ)
    exact instLowerTopologyProd (α := αᵒᵈ) (β := βᵒᵈ)
```
This may be a bit too much of defeq abuse, but I don't see any other solution that doesn't involve a ton of extra API. @YaelDillies could I get your opinion on this?

Reply:
I've committed this for now, but happy to take an alternative suggestion from @YaelDillies .

Reply:
That `(α × β)ᵒᵈ ≃o αᵒᵈ × βᵒᵈ` is something I've had my eyes on for a while. I never had to use it yet, so I'm pleasantly surprised to see it being so painless.

I honestly think your approach is fine, but let me spell out the alternative as there's a few ingredients you could add already:
* define a type `α ≃ₜo β` of topological order isomorphisms
* show `(α × β)ᵒᵈ ≃ₜo αᵒᵈ × βᵒᵈ`. This is trivial, eg here's the one for order isomorphisms: `def Prod.OrderDualDistrib (α β : Type _) [Preorder α] [Preorder β] : (α × β)ᵒᵈ ≃o αᵒᵈ × βᵒᵈ :=
OrderIso.refl _` (this one you could add already)
* show that `UpperTopology` is preserved under topological order isomorphisms. Should be trivial as well.
* show that `UpperTopology αᵒᵈ ↔ LowerTopology α` (this you could add already)

Reply:
Yes it doesn't look super bad, but somehow it's even more worrying because I can't explain why it works so well :sweat_smile: 
What you describe is indeed what I would have suggested as a "clean" approach, but it deserves its own PR. I'll open an issue for it. In the meantime, @mans0954 could you prove `UpperTopology αᵒᵈ ↔ LowerTopology α` as well as `LowerTopology αᵒᵈ ↔ UpperTopology α` and make the backwards direction an instance? It's already the case for one of them, but since `(αᵒᵈ)ᵒᵈ` is not the same as `α` for the typeclass system you can add the other one without any troubles.

Reply:
@ADedecker is this what you had in mind? https://github.com/leanprover-community/mathlib4/pull/5873/commits/f9ef40a79dbc489831d902d24ad68621eaf3a64a

Reply:
Yes, thanks!